python/common/org/python/types/List.java

```diff
@@ -167,8 +168,14 @@ public List(org.python.Object [] args, java.util.Map<java.lang.String, org.pytho
             int count = Math.min(size, otherSize);
 
             for (int i = 0; i < count; i++) {
-                org.python.types.Bool b =
-                    (org.python.types.Bool) this.value.get(i).__lt__(otherList.value.get(i));
+                org.python.Object result = this.value.get(i).__lt__(otherList.value.get(i));
```

Comment:
(Also, there should probably be a `PyObject_RichCompareBool` which does some extra checks.

```
>>> class F:
...     def __init__(self, v):
...         self.v = v
...     def __eq__(self, other):
...         return self.v == other.v + 1
...
>>> k = F(5)
>>> k == k
False
>>> [k] == [k]
True
>>> {'hi': k} == {'hi': k}
True
```

Looking at the Python source code, it seems the "normal" compare (`COMPARE_OP`) goes to `PyObject_RichCompare`, while containers (list/dict) use `PyObject_RichCompareBool`). Of course that's not a precise argument, but it does indicate that maybe there should be several methods.